runtime/fetcher/fetchasstream.go

```diff
@@ -0,0 +1,104 @@
+package fetcher
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+)
+
+// ErrStreamReset returned from io.Reader when the fetch process is reset
+var ErrStreamReset = errors.New("stream was reset by fetcher")
+
+// A StreamHandler is a function that handles a stream, the stream maybe a
+// aborted in which case io.Reader will return ErrStreamReset and the Context
+// will be canceled.
+type StreamHandler func(context.Context, io.Reader) error
+
+// FetchAsStream gets a reference as a stream.
+//
+// Notice that target StreamHandler may be invoked multiple times, if the
+// connection breaks while fetching it might be retried. In which case the
+// Context passed to the target StreamHandler will be canceled.
+func FetchAsStream(context Context, reference Reference, target StreamHandler) error {
```

Comment:
IIUC, we will call the `StreamHandler` only if `Reset` is called, which may never happen if there is no error.

Reply:
The `reference.Fetch` may call `streamReseter.Reset`, I'll add a comment to highlight this, because that would be useful :)